solvebio/cli/data.py

```diff
@@ -505,6 +506,113 @@ def _download(full_path, local_folder_path, dry_run=False):
+        # TODO: Skip dataset objects for now?
+        if object_.is_dataset:
+            continue
+        # add local download path
+        object_.local_path = os.path.join(local_folder_path, object_.filename)
+        remote_files.append(object_)
+        if object_.is_folder:
+            _get_remote_directory_recursive(
+                    full_path=object_.full_path + "/*",
+                    local_folder_path=object_.local_path,
+                    remote_files=remote_files)
+    return remote_files
+
+
+def _get_local_directory_recursive(local_folder_path):
+    """Queries local directory to find all files"""
+    local_files = []
+    for root, dirs, files in os.walk(local_folder_path):
+        for file_ in files:
+            path = os.path.join(root, file_)
+            md5 = md5sum(path)
+            local_files.append(dict(
+                path=os.path.join(root, file_), file_type="file", md5=md5))
+        for dir_ in dirs:
+            local_files.append(dict(
+                path=os.path.join(root, dir_), file_type="folder"))
+    return local_files
+
+
+
+
+def _sync(full_path, local_folder_path, dry_run=False, delete_local=False):
+    print("Syncing from {} to {}".format(
+            full_path, local_folder_path))
+
+    if dry_run:
+        print('Running in dry run mode. Not downloading any files.')
+
+    local_folder_path = os.path.expanduser(local_folder_path)
+    if not os.path.exists(local_folder_path):
+        print("Creating local download folder {}".format(local_folder_path))
+        if not dry_run:
+            if not os.path.exists(local_folder_path):
+                os.makedirs(local_folder_path)
+
+    # Get all remote files
+    remote_files = _get_remote_directory_recursive(full_path, local_folder_path)
+    for file_ in remote_files:
+        print(file_)
+
+    # Map remote files to their local file path
+    local_file_map = {x.local_path:dict(remote=x) for x in remote_files}
+    local_files = _get_local_directory_recursive(local_folder_path)
+    for local_file in local_files:
+        # For each local file, add to the map
+        local_path = local_file['path']
+        if local_path in local_file_map:
+            local_file_map[local_path]['local'] = local_file
+        else:
+            # Delete any files not necessary?
+            """
+            if delete and not dry_run:
+                os.unlink(local_file)
+            """
+            pass
+
+    # Loop through the map of local file paths
+    # and identify:
+    # - directories to create
+    # - files to download
+    updates = []
+    for local_path, file_obj_map in local_file_map.items():
+        print(file_obj_map)
+        remote_file = file_obj_map['remote']
+
+        if remote_file.object_type == "folder":
+            os.makedirs(local_path, exist_ok=True)
+            continue
+
+        if "local" in file_obj_map:
+            if file_obj_map['remote']['md5'] != file_obj_map['local']['md5']:
```

Comment:
It "might" be possible for some remote files to not have md5. In this case, just set the remote md5 to None and assume it is always different (which might already be happening here)

Reply:
Will do. I _think_ right now it would just fail via a KeyError if `md5` didn't exist? I'm not sure if the server actually returns `md5 = null` or just omits the key. I'll take a closer look and adjust accordingly.